docs/guides/virtual-tables.md

```diff
@@ -0,0 +1,165 @@
+They are much more lightweight than their regular counterparts and they completely lack all the burden and overhead of updating a persistent storage and keeping it consistent with the already existing in-memory structures.
+Instead virtual table readers can just translate these in-memory structures into CQL results at the time they are queried.
+These tables completely look and act like regular tables from a user's point of view, which is the entire point: the user can use an interface they are already familiar with to query and manipulate data.
+
+
+## What is a good candidate for a virtual table?
+
+In general a good candidate for a virtual table has to match the following conditions:
+1) It exposes data readily available in in-memory structures of the node. The data doesn't have to be related to the node, it can be about the entire cluster too.
+2) The amount of data is small to moderate (partitions/rows in the order of low thousands).
+3) The data is not available from other tables, like regular system tables.
+
+
+## Information retrieval nodetool commands as virtual tables
+
+Some time ago we started adding virtual tables to replace information retrieval nodetool commands.
+These nodetool commands are currently served by one or more REST API endpoints.
+CQL tables are a natural fit to replace these nodetool commands:
+* One can retrieve the information with just a CQL client, instead of having to have both jmx and nodetool setup;
+* Remote access: REST and therefore nodetool is only accessible on the same machine the node lives on for security considerations (see below);
+* Security: CQL has authentication and authorization already, with fine-grained RBAC support;
+* Filtering and aggregation: CQL allows for filtering and selecting rows/columns as well as aggregating them;
+
+Even though not widely known, CQL also has built-in JSON support (`select json...`) so if somebody prefers JSON (used by REST) they can still use JSON with CQL too.
+
+Another big advantage to using CQL virtual tables is that we have a free hand at what to include and how to format it, while the nodetool interface is bound (to some degree) by our drop-in replacement promise: users coming from Apache Cassandra expectheir familiar nodetool interface.
+
+
+## How to add a new virtual table?
+
+The process of adding a new virtual table is as follows:
+* Choose the appropriate class to inherit from: `db::memtable_filling_virtual_table` or `db::streaming_virtual_table` (located in `db/virtual_table.hh`);
+* Implement the interface generating the data, mind shard awareness and query restrictions if they apply;
+* Instantiate and register your virtual table in `register_virtual_tables()` in `db/system_keyspace.cc`;
+
+### Choosing the right class for you virtual table
+
+#### memtable_filling_virtual_table
+
+If your table generates a fixed and small amount of data you should use the `db::memtable_filling_virtual_table`.
```

Comment:
Right, it can cause quadratic effort, since in the worst case every page needs the entire memtable filled again.

Reply:
Synchoronos vs. asynchronos doesn't matter. I have a table in this PR (`system.runtime_info`) that has lots of `co_await` in it but still uses the memtable because it has a known and constant amount of items.